### PR TITLE
Allow logstream to stream logs from user namespace

### PR DESCRIPTION
**Pull request by mgencur**
* When streaming logs from user namespace, disable filtering lines so all lines are printed.

* More information on when this comes handy is available in Serving PR: https://github.com/knative/serving/pull/13293, together with usage example.
It will be very handy also in knative/eventing upgrade tests where multiple pods are started and the Sender pod is removed before the test ends.

# Changes

-
-
-

/kind <kind>

Fixes #

**Release Note**

```release-note

```

**Docs**

```docs

```

### CODE REVIEW

**Comment by mgencur** [comment]
The test failures look unrelated.

File: test/logstream/interface.go [hunk 1]
```diff
@@ -31,7 +32,10 @@ import (
 
 // Canceler is the type of a function returned when a logstream is started to be
 // deferred so that the logstream can be stopped when the test is complete.
-type Canceler = logstreamv2.Canceler
+type (
+	Canceler = logstreamv2.Canceler
+	Callback = logstreamv2.Callback
```
Comment (skonto) [inline test/logstream/interface.go:37]: I guess we need to document `Callback` too here.

Comment (mgencur) [reply test/logstream/interface.go:37]: Oh, I see.

Comment (mgencur) [reply test/logstream/interface.go:38]: Pushed.

File: test/logstream/v2/stream.go [hunk 2]
```diff
@@ -34,21 +34,25 @@ import (
...
+		kc:          c,
+		namespaces:  namespaces,
+		keys:        make(map[string]Callback, 1),
+		filterLines: filterLines,
+		podPrefixes: podPrefixes,
 	}
 }
 
-func FromNamespace(ctx context.Context, c kubernetes.Interface, namespace string) Source {
+func FromNamespace(ctx context.Context, c kubernetes.Interface, namespace string, filterLines bool, podPrefixes ...string) Source {
```
Comment (dprotaso) [inline test/logstream/v2/stream.go:48]: This is a breaking API change. You could extend this method without breaking by using variadic functional options

see option 3 - https://golang.cafe/blog/golang-functional-options-pattern.html

File: test/logstream/v2/stream.go [hunk 3]
```diff
@@ -34,21 +34,25 @@ import (
 	"knative.dev/pkg/ptr"
 )
 
-func FromNamespaces(ctx context.Context, c kubernetes.Interface, namespaces []string) Source {
+func FromNamespaces(ctx context.Context, c kubernetes.Interface, namespaces []string, filterLines bool, podPrefixes ...string) Source {
```
Comment (dprotaso) [inline test/logstream/v2/stream.go:37]: I don't have a good suggestion on what to do here that's not a breaking change.

File: test/logstream/interface.go [hunk 4]
```diff
@@ -41,53 +46,69 @@ type ti interface {
...
 
-	return stream.Start(t)
+	return sysStream.Start(t, t.Logf)
+}
+
+// StartForUserNamespace begins streaming the logs from custom namespaces.
+// Filtering of log lines is disabled in this case so all lines will be printed
+// through the provided callback.
+// It returns a Canceler which must be called before the test completes.
+func StartForUserNamespace(t ti, callback Callback, namespace string, podPrefixes ...string) Canceler {
```
[the remark this answers is not in the document]
Comment (mgencur) [reply test/logstream/interface.go]: OK. Good to know.

File: test/logstream/v2/stream.go (hunk 3 above)
Comment (mgencur) [reply test/logstream/v2/stream.go]: This is addressed in the latest commit.

File: test/logstream/v2/stream.go (hunk 2 above)
Comment (mgencur) [reply test/logstream/v2/stream.go]: Thanks. I have tried what you're suggesting and pushed a new commit. Please let me know what you think. I think this new way of constructing the logstream could replace the "FromNamespace" and "FromNamespaces" functions so we could mark them as deprecated (which I haven't done yet).

Comment (mgencur) [reply test/logstream/v2/stream.go]: I have also vendored the changes into https://github.com/knative/serving/pull/13293 so the exaple usage can be seen there.

**Comment by mgencur** [comment]
Rebased. Just to be sure this works after some time.

@dprotaso Could you review this please? There's a commit which haven't been reviewed. See [comment](https://github.com/knative/pkg/pull/2591#discussion_r973972854)

**Comment by mgencur** [comment]
@dprotaso Thank you!

**Pull request merged**